File: Answer/396.旋转函数.go
/*
 * @lc app=leetcode.cn id=396 lang=golang
 *
 * [396] 旋转函数
 */

// @lc code=start
func maxRotateFunction(nums []int) int {
	n := len(nums)

	sum, f := 0, 0
	for i, v := range nums {
		sum += v
		f += i * v
	}
	
	maxF := f
	for i := n - 1; i > 0; i-- {
		f += sum - n*nums[i]
		if f > maxF {
			maxF = f
		}
	}
	return maxF
}

// @lc code=end

File: Answer/396.旋转函数.java
/*
 * @lc app=leetcode.cn id=396 lang=java
 *
 * [396] 旋转函数
 */

// @lc code=start
class Solution {
    public int maxRotateFunction(int[] nums) {
        int n = nums.length;

        int sum = 0, f = 0;
        for (int i = 0; i < n; i++) {
            sum += nums[i];
            f += i * nums[i];
        }

        int maxF = f;
        for (int i = n - 1; i > 0; i--) {
            f += sum - n * nums[i];
            if (f > maxF) {
                maxF = f;
            }
        }
        return maxF;
    }
}
// @lc code=end
